src/pip/_internal/resolution/resolvelib/requirements.py

```diff
@@ -33,6 +33,27 @@ def is_satisfied_by(self, candidate):
         return candidate == self.candidate
 
 
+class NoMatchRequirement(Requirement):
+    """A requirement that never matches anything.
+    """
+    def __init__(self, name):
+        # type: (str) -> None
+        self._name = canonicalize_name(name)
```

Comment:
Is the canonicalization necessary here? Wouldn't the only value this sees be `<Python>`?

Reply:
You’re right. I’ll also change the argument name to `canonical_name` to make this more obvious.

Reply:
Given that `name` is only used for the resolver's `identify` method, and hence to disambiguate nodes in the resolution graph, there's no requirement that names be canonicalised for "artificial" candidates/requirements like this.

I'd suggest we leave it not canonicalised, it means there's even less risk of clashing with a PyPI package...

Reply:
Fixed. I also added a note in the class docstring to explain the expectation.